File: src/main/java/com/khmelyuk/multirun/MultirunProcessRegistry.java
package com.khmelyuk.multirun;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.intellij.execution.process.BaseProcessHandler;
import com.intellij.execution.process.ProcessEvent;
import com.intellij.execution.process.ProcessHandler;
import com.intellij.execution.process.ProcessListener;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.openapi.project.Project;

/**
 * Keeps track of every process started by a Multiple Run configuration, so the
 * "Multiple Run Monitor" tool window can list them with live memory statistics.
 * Entries are removed automatically when the process terminates.
 */
public final class MultirunProcessRegistry {

    /** One running application started by a Multirun configuration. */
    public static final class Entry {
        public final String multirunName;
        public final String appName;
        public final ProcessHandler handler;
        /** Configured memory cap in MB, or null when the application has no limit. */
        public final Integer memoryLimitMb;
        /** The environment the app was launched with; lets the monitor restart just this app. */
        public final ExecutionEnvironment environment;
        /** File name of the active env profile at launch time, or "-" when none. */
        public final String envFileName;
        /** Raw "Ready when" condition of the app (port:/log:/http...), or null when none. */
        public final String readyCondition;
        public final long startedAtMs;

        Entry(String multirunName, String appName, ProcessHandler handler,
              Integer memoryLimitMb, ExecutionEnvironment environment, String envFileName,
              String readyCondition) {
            this.multirunName = multirunName;
            this.appName = appName;
            this.handler = handler;
            this.memoryLimitMb = memoryLimitMb;
            this.environment = environment;
            this.envFileName = envFileName == null || envFileName.isEmpty() ? "-" : envFileName;
            this.readyCondition = readyCondition;
            this.startedAtMs = System.currentTimeMillis();
        }
    }

    private static final Map<Project, List<Entry>> ENTRIES = new ConcurrentHashMap<>();

    private MultirunProcessRegistry() {
    }

    public static void register(@NotNull Project project, String multirunName, String appName,
                                @NotNull ProcessHandler handler, @Nullable Integer memoryLimitMb,
                                @Nullable ExecutionEnvironment environment, @Nullable String envFileName,
                                @Nullable String readyCondition) {
        final Entry entry = new Entry(multirunName, appName, handler, memoryLimitMb, environment,
                                      envFileName, readyCondition);
        ENTRIES.computeIfAbsent(project, p -> new CopyOnWriteArrayList<>()).add(entry);
        handler.addProcessListener(new ProcessListener() {
            @Override
            public void processTerminated(@NotNull ProcessEvent event) {
                unregister(project, handler);
            }
        });
        // the handler may have died between the run and this call
        if (handler.isProcessTerminated()) {
            unregister(project, handler);
        }
        MemoryLimitWatcher.ensureStarted();
    }

    public static void unregister(@NotNull Project project, @NotNull ProcessHandler handler) {
        final List<Entry> list = ENTRIES.get(project);
        if (list != null) {
            list.removeIf(entry -> entry.handler == handler);
        }
    }

    @NotNull
    public static List<Entry> getEntries(@NotNull Project project) {
        final List<Entry> list = ENTRIES.get(project);
        return list == null ? Collections.emptyList() : new ArrayList<>(list);
    }

    /** A copy of all live entries of all projects; used by the background memory limit watcher. */
    @NotNull
    public static Map<Project, List<Entry>> snapshot() {
        final Map<Project, List<Entry>> copy = new LinkedHashMap<>();
        ENTRIES.forEach((project, entries) -> {
            if (!entries.isEmpty()) {
                copy.put(project, new ArrayList<>(entries));
            }
        });
        return copy;
    }

    /** The OS pid behind the handler, or -1 when it cannot be determined. */
    public static long pidOf(@NotNull ProcessHandler handler) {
        if (handler instanceof BaseProcessHandler) {
            try {
                return ((BaseProcessHandler<?>) handler).getProcess().pid();
            } catch (UnsupportedOperationException ignored) {
                // some Process implementations don't expose a pid
            }
        }
        return -1;
    }
}

File: src/main/kotlin/io/github/welingtonmonteiro/multiplerun/MultirunProcessRegistry.kt
package io.github.welingtonmonteiro.multiplerun

import java.util.Collections
import java.util.LinkedHashMap
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

import com.intellij.execution.process.BaseProcessHandler
import com.intellij.execution.process.ProcessEvent
import com.intellij.execution.process.ProcessHandler
import com.intellij.execution.process.ProcessListener
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.openapi.project.Project

/**
 * Keeps track of every process started by a Multiple Run configuration, so the
 * "Multiple Run Monitor" tool window can list them with live memory statistics.
 * Entries are removed automatically when the process terminates.
 */
class MultirunProcessRegistry private constructor() {

    /** One running application started by a Multirun configuration. */
    class Entry internal constructor(
        @JvmField val multirunName: String,
        @JvmField val appName: String,
        @JvmField val handler: ProcessHandler,
        /** Configured memory cap in MB, or null when the application has no limit. */
        @JvmField val memoryLimitMb: Int?,
        /** The environment the app was launched with; lets the monitor restart just this app. */
        @JvmField val environment: ExecutionEnvironment?,
        rawEnvFileName: String?,
        rawLoadedEnv: Map<String, String>?,
        /** Whether the app also inherits the system/parent environment on top of [loadedEnv]. */
        @JvmField val includeSystemEnv: Boolean,
        /** Raw "Ready when" condition of the app (port:/log:/http...), or null when none. */
        @JvmField val readyCondition: String?,
        rawMemAlertThreshold: Int,
        /** true = restart the app at the threshold; false = just notify. */
        @JvmField val memLimitRestart: Boolean,
        rawCpuAlertThreshold: Int,
    ) {
        /** File name of the active env profile at launch time, or "-" when none. */
        @JvmField
        val envFileName: String = if (rawEnvFileName.isNullOrEmpty()) "-" else rawEnvFileName

        /**
         * The environment variables Multiple Run actually injected into this app at launch
         * (group variables + group env file + memory-limit options + per-app env file, merged),
         * so the monitor can show exactly what was loaded. Never logged - values stay in memory.
         */
        @JvmField
        val loadedEnv: Map<String, String> = if (rawLoadedEnv == null) emptyMap()
            else Collections.unmodifiableMap(LinkedHashMap(rawLoadedEnv))

        /** Percent of the memory limit that triggers the alert/action. */
        @JvmField
        val memAlertThreshold: Int = if (rawMemAlertThreshold <= 0) 90 else rawMemAlertThreshold

        /** Sustained CPU % that triggers an alert for this app, or 0 when disabled. */
        @JvmField
        val cpuAlertThreshold: Int = maxOf(0, rawCpuAlertThreshold)

        @JvmField
        val startedAtMs: Long = System.currentTimeMillis()
    }

    companion object {
        private val ENTRIES = ConcurrentHashMap<Project, MutableList<Entry>>()

        /**
         * Last multirun launch metadata per app name. Unlike ENTRIES this survives process
         * termination, so an app restarted individually (outside the multirun umbrella) still
         * shows its group, env profile and memory limit in the monitor.
         */
        private val LAST_BY_NAME = ConcurrentHashMap<Project, ConcurrentHashMap<String, Entry>>()

        @JvmStatic
        fun register(
            project: Project, multirunName: String, appName: String,
            handler: ProcessHandler, memoryLimitMb: Int?,
            environment: ExecutionEnvironment?, envFileName: String?,
            loadedEnv: Map<String, String>?, includeSystemEnv: Boolean,
            readyCondition: String?, memAlertThreshold: Int, memLimitRestart: Boolean,
            cpuAlertThreshold: Int,
        ) {
            val entry = Entry(multirunName, appName, handler, memoryLimitMb, environment,
                              envFileName, loadedEnv, includeSystemEnv, readyCondition,
                              memAlertThreshold, memLimitRestart, cpuAlertThreshold)
            ENTRIES.computeIfAbsent(project) { CopyOnWriteArrayList() }.add(entry)
            LAST_BY_NAME.computeIfAbsent(project) { ConcurrentHashMap() }.put(appName, entry)
            handler.addProcessListener(object : ProcessListener {
                override fun processTerminated(event: ProcessEvent) {
                    unregister(project, handler)
                }
            })
            // the handler may have died between the run and this call
            if (handler.isProcessTerminated()) {
                unregister(project, handler)
            }
            MemoryLimitWatcher.ensureStarted()
        }

        @JvmStatic
        fun unregister(project: Project, handler: ProcessHandler) {
            ENTRIES[project]?.removeAll { it.handler === handler }
        }

        @JvmStatic
        fun getEntries(project: Project): List<Entry> {
            val list = ENTRIES[project]
            return if (list == null) emptyList() else ArrayList(list)
        }

        /** A copy of all live entries of all projects; used by the background memory limit watcher. */
        @JvmStatic
        fun snapshot(): Map<Project, List<Entry>> {
            val copy = LinkedHashMap<Project, List<Entry>>()
            ENTRIES.forEach { project, entries ->
                if (entries.isNotEmpty()) {
                    copy[project] = ArrayList(entries)
                }
            }
            return copy
        }

        /**
         * Metadata of the last multirun launch of an app with this name, or null when the app was
         * never started by Multiple Run. The entry's handler/startedAt may belong to a dead process -
         * callers must only use the descriptive fields (group, limit, env profile, condition).
         */
        @JvmStatic
        fun findMetadataByName(project: Project, appName: String?): Entry? {
            val byName = LAST_BY_NAME[project]
            return if (byName == null || appName == null) null else byName[appName]
        }

        /** The OS pid behind the handler, or -1 when it cannot be determined. */
        @JvmStatic
        fun pidOf(handler: ProcessHandler): Long {
            if (handler is BaseProcessHandler<*>) {
                try {
                    return handler.getProcess().pid()
                } catch (ignored: UnsupportedOperationException) {
                    // some Process implementations don't expose a pid
                }
            }
            return -1
        }
    }
}
